src/common/RefCountedObj.h

```diff
@@ -66,30 +69,128 @@ struct RefCountedObject {
       ANNOTATE_HAPPENS_BEFORE(&nref);
     }
   }
-  void set_cct(CephContext *c) {
-    cct = c;
+
+protected:
+  RefCountedObject() = default;
+  RefCountedObject(CephContext* c) : cct(c) {}
+
+  virtual ~RefCountedObject() {
+    ceph_assert(nref == 0);
   }
 
-  uint64_t get_nref() const {
-    return nref;
+private:
+  mutable std::atomic<uint64_t> nref = {1};
```

Comment:
unrelated, but i wonder if we should make this int64_t so that we don't wrap to 0xffffffffffffffff with too many puts

Reply:
Sure, I'll change it.